File: lib/src/core/access/role_access.dart
import '../../models/user.dart';
import 'module_registry.dart';

/// Aturan hak akses per peran — port dari `apps/frontend/lib/role-access.ts`.
///
/// Konfigurasi tersimpan lokal (per perangkat) dan hanya membatasi navigasi di
/// sisi klien. Otorisasi sebenarnya tetap ditegakkan backend lewat
/// `requireRole`, jadi ini murni lapisan pengalaman pengguna.
class RoleAccess {
  const RoleAccess._();

  /// Peran yang dapat dikonfigurasi di layar pengaturan hak akses.
  static const List<UserRole> managedRoleOrder = [
    UserRole.admin,
    UserRole.dokter,
    UserRole.perawat,
    UserRole.bidan,
    UserRole.teknis,
    UserRole.umum,
  ];

  /// Rute yang boleh diakses tanpa sesi.
  static const Set<String> publicPaths = {
    '/',
    '/login',
    '/daftar',
    '/lupa-password',
  };

  /// Peran yang boleh membaca tiap koleksi, dipakai untuk menyembunyikan menu
  /// yang pasti ditolak backend. Cerminan `READABLE_COLLECTION_ROLES`.
  static const Map<String, List<UserRole>> readableCollectionRoles = {
    'patients': [
      UserRole.admin,
      UserRole.dokter,
      UserRole.bidan,
      UserRole.perawat,
      UserRole.teknis,
      UserRole.umum,
    ],
    'users': [UserRole.admin],
    'doctors': [
      UserRole.admin,
      UserRole.dokter,
      UserRole.bidan,
      UserRole.perawat,
      UserRole.teknis,
      UserRole.umum,
    ],
    'services': [
      UserRole.admin,
      UserRole.dokter,
      UserRole.bidan,
      UserRole.perawat,
      UserRole.teknis,
      UserRole.umum,
    ],
    'appointments': [
      UserRole.admin,
      UserRole.dokter,
      UserRole.bidan,
      UserRole.perawat,
      UserRole.umum,
    ],
    'medical-records': [
      UserRole.admin,
      UserRole.dokter,
      UserRole.bidan,
      UserRole.perawat,
      UserRole.teknis,
      UserRole.umum,
    ],
    'medicines': [
      UserRole.admin,
      UserRole.teknis,
      UserRole.dokter,
      UserRole.bidan,
      UserRole.perawat,
      UserRole.umum,
    ],
    'medical-equipments': [
      UserRole.admin,
      UserRole.perawat,
      UserRole.dokter,
      UserRole.bidan,
    ],
    'lab-orders': [
      UserRole.admin,
      UserRole.teknis,
      UserRole.dokter,
      UserRole.bidan,
      UserRole.perawat,
    ],
    'lab-results': [
      UserRole.admin,
      UserRole.teknis,
      UserRole.dokter,
      UserRole.bidan,
      UserRole.perawat,
    ],
    'radiology-orders': [
      UserRole.admin,
      UserRole.teknis,
      UserRole.dokter,
      UserRole.bidan,
      UserRole.perawat,
    ],
    'payments': [
      UserRole.admin,
      UserRole.dokter,
      UserRole.bidan,
      UserRole.perawat,
      UserRole.teknis,
      UserRole.umum,
    ],
    'billing-records': [
      UserRole.admin,
      UserRole.dokter,
      UserRole.bidan,
      UserRole.perawat,
      UserRole.teknis,
      UserRole.umum,
    ],
    'documents': [UserRole.admin, UserRole.umum],
    'pharmacy-requests': [
      UserRole.admin,
      UserRole.teknis,
      UserRole.dokter,
      UserRole.bidan,
      UserRole.perawat,
      UserRole.umum,
    ],
    'insurance-profiles': [
      UserRole.admin,
      UserRole.umum,
      UserRole.dokter,
      UserRole.bidan,
    ],
    'insurance-bridge-members': [UserRole.admin, UserRole.umum],
    'beds': [UserRole.admin, UserRole.perawat, UserRole.dokter, UserRole.bidan],
    'inpatient-admissions': [
      UserRole.admin,
      UserRole.perawat,
      UserRole.dokter,
      UserRole.bidan,
    ],
    'doctor-visit-notes': [
      UserRole.admin,
      UserRole.dokter,
      UserRole.bidan,
      UserRole.perawat,
    ],
    'patient-notifications': [UserRole.admin, UserRole.umum],
    'satisfaction-surveys': [UserRole.admin, UserRole.umum],
    'audit-logs': [UserRole.admin],
    'referrals': [
      UserRole.admin,
      UserRole.dokter,
      UserRole.bidan,
      UserRole.perawat,
      UserRole.umum,
    ],
    'referral-facilities': [
      UserRole.admin,
      UserRole.dokter,
      UserRole.bidan,
      UserRole.perawat,
      UserRole.umum,
    ],
  };

  /// Konfigurasi bawaan: setiap peran terkelola mendapat seluruh modul, lalu
  /// dapat dipersempit administrator lewat layar pengaturan.
  static Map<UserRole, List<String>> defaultSettings() => {
        for (final role in managedRoleOrder) role: List.of(allModuleHrefs),
      };

  /// Membersihkan konfigurasi tersimpan: membuang href yang sudah tidak ada
  /// dan memastikan admin selalu memegang akses penuh.
  static Map<UserRole, List<String>> normalize(
    Map<UserRole, List<String>> settings,
  ) {
    final defaults = defaultSettings();
    final validHrefs = allModuleHrefs.toSet();

    return {
      for (final role in managedRoleOrder)
        role: role == UserRole.admin
            ? defaults[role]!
            : (settings[role] ?? defaults[role]!)
                .toSet()
                .where(validHrefs.contains)
                .toList(),
    };
  }

  /// Mengubah bentuk tersimpan (kunci string) menjadi peta ber-enum.
  static Map<UserRole, List<String>> fromStorage(
    Map<String, List<String>> raw,
  ) {
    if (raw.isEmpty) return defaultSettings();

    final parsed = <UserRole, List<String>>{};
    raw.forEach((key, value) {
      final role = UserRole.tryParse(key);
      if (role != null) parsed[role] = value;
    });

    return normalize(parsed);
  }

  static Map<String, List<String>> toStorage(
    Map<UserRole, List<String>> settings,
  ) =>
      settings.map((role, hrefs) => MapEntry(role.value, hrefs));

  static Set<String> allowedModuleHrefs(
    UserRole? role,
    Map<UserRole, List<String>> settings,
  ) {
    if (role == null) return const {};
    return normalize(settings)[role]?.toSet() ?? const {};
  }

  /// Apakah peran boleh membuka sebuah rute.
  ///
  /// Rute yang tidak terdaftar sebagai modul (mis. sub-halaman detail)
  /// dicocokkan ke modul induk terpanjang yang menjadi prefiksnya.
  static bool canAccessPath(
    UserRole? role,
    String path,
    Map<UserRole, List<String>> settings,
  ) {
    if (publicPaths.contains(path)) return true;
    if (role == null) return false;

    final matched = allModuleHrefs
        .where((href) => path == href || path.startsWith('$href/'))
        .fold<String?>(
          null,
          (longest, href) =>
              longest == null || href.length > longest.length ? href : longest,
        );

    if (matched == null) return true;
    return allowedModuleHrefs(role, settings).contains(matched);
  }

  /// Rute tujuan setelah login: dashboard bila diizinkan, selain itu modul
  /// pertama yang boleh diakses.
  static String landingPath(
    UserRole? role,
    Map<UserRole, List<String>> settings,
  ) {
    if (role == null) return '/login';

    final allowed = allowedModuleHrefs(role, settings);
    if (allowed.contains('/dashboard')) return '/dashboard';

    return allModuleHrefs.firstWhere(
      allowed.contains,
      orElse: () => '/pengaturan',
    );
  }

  static bool canReadCollection(UserRole? role, String collection) {
    if (role == null) return false;
    if (role == UserRole.admin) return true;

    final base = collection.startsWith('documents?category=')
        ? 'documents'
        : collection;
    return readableCollectionRoles[base]?.contains(role) ?? false;
  }
}

File: apps/frontend/lib/role-access.ts
import type { User } from "@/lib/auth-types";
import { moduleSections, type UserRole } from "@/lib/module-registry";

export const ROLE_ACCESS_STORAGE_KEY = "clinic_role_access_settings_v2";
export const ROLE_ACCESS_CHANGED_EVENT = "clinic-role-access-changed";

export type RoleAccessSettings = Partial<Record<UserRole, string[]>>;

export const MANAGED_ROLE_ORDER: UserRole[] = [
  "admin",
  "dokter",
  "perawat",
  "bidan",
  "teknis",
  "umum",
];

const allModuleHrefs = moduleSections.flatMap((section) => section.items.map((item) => item.href));
const validModuleHrefs = new Set(allModuleHrefs);
const moduleByHref = new Map(
  moduleSections.flatMap((section) => section.items.map((item) => [item.href, item]))
);
const PUBLIC_PATHS = new Set(["/", "/login", "/daftar", "/lupa-password"]);

const READABLE_COLLECTION_ROLES: Record<string, UserRole[]> = {
  patients: ["admin", "dokter", "bidan", "perawat", "teknis", "umum"],
  users: ["admin"],
  doctors: ["admin", "dokter", "bidan", "perawat", "teknis", "umum"],
  services: ["admin", "dokter", "bidan", "perawat", "teknis", "umum"],
  appointments: ["admin", "dokter", "bidan", "perawat", "umum"],
  "appointments/today": ["admin", "dokter", "bidan", "perawat", "umum"],
  "medical-records": ["admin", "dokter", "bidan", "perawat", "teknis", "umum"],
  medicines: ["admin", "teknis", "dokter", "bidan", "perawat", "umum"],
  "medical-equipments": ["admin", "perawat", "dokter", "bidan"],
  "lab-orders": ["admin", "teknis", "dokter", "bidan", "perawat"],
  "lab-results": ["admin", "teknis", "dokter", "bidan", "perawat"],
  "radiology-orders": ["admin", "teknis", "dokter", "bidan", "perawat"],
  payments: ["admin", "dokter", "bidan", "perawat", "teknis", "umum"],
  "billing-records": ["admin", "dokter", "bidan", "perawat", "teknis", "umum"],
  "payment-summaries": ["admin", "dokter", "bidan", "perawat", "teknis", "umum"],
  documents: ["admin", "umum"],
  "pharmacy-requests": ["admin", "teknis", "dokter", "bidan", "perawat", "umum"],
  "insurance-profiles": ["admin", "umum", "dokter", "bidan"],
  "insurance-bridge-members": ["admin", "umum"],
  beds: ["admin", "perawat", "dokter", "bidan"],
  "inpatient-admissions": ["admin", "perawat", "dokter", "bidan"],
  "doctor-visit-notes": ["admin", "dokter", "bidan", "perawat"],
  "patient-notifications": ["admin", "umum"],
  "satisfaction-surveys": ["admin", "umum"],
  "audit-logs": ["admin"],
  referrals: ["admin", "dokter", "bidan", "perawat", "umum"],
  "referral-facilities": ["admin", "dokter", "bidan", "perawat", "umum"],
};

function canRoleUseModule(role: UserRole, href: string) {
  const module = moduleByHref.get(href);
  return Boolean(module && role);
}

export function getAllAccessModules() {
  return moduleSections.flatMap((section) =>
    section.items.map((item) => ({
      sectionId: section.id,
      sectionLabel: section.label,
      title: item.title,
      shortTitle: item.shortTitle,
      href: item.href,
      icon: item.icon,
    }))
  );
}

export function getDefaultRoleAccessSettings(): Record<UserRole, string[]> {
  const defaults = {} as Record<UserRole, string[]>;

  MANAGED_ROLE_ORDER.forEach((role) => {
    defaults[role] = allModuleHrefs.filter((href) => canRoleUseModule(role, href));
  });

  return defaults;
}

function normalizeRoleAccessSettings(settings: RoleAccessSettings): Record<UserRole, string[]> {
  const defaults = getDefaultRoleAccessSettings();

  return Object.fromEntries(
    (Object.keys(defaults) as UserRole[]).map((role) => {
      if (role === "admin") return [role, defaults[role]];

      const configuredHrefs = settings[role];
      if (!configuredHrefs) return [role, defaults[role]];

      return [
        role,
        Array.from(new Set(configuredHrefs)).filter(
          (href) => validModuleHrefs.has(href) && canRoleUseModule(role, href)
        ),
      ];
    })
  ) as Record<UserRole, string[]>;
}

export function loadRoleAccessSettings(): Record<UserRole, string[]> {
  if (typeof window === "undefined") return getDefaultRoleAccessSettings();

  const rawSettings = window.localStorage.getItem(ROLE_ACCESS_STORAGE_KEY);
  if (!rawSettings) return getDefaultRoleAccessSettings();

  try {
    return normalizeRoleAccessSettings(JSON.parse(rawSettings) as RoleAccessSettings);
  } catch {
    return getDefaultRoleAccessSettings();
  }
}

export function saveRoleAccessSettings(settings: RoleAccessSettings): Record<UserRole, string[]> {
  const normalizedSettings = normalizeRoleAccessSettings(settings);

  if (typeof window !== "undefined") {
    window.localStorage.setItem(ROLE_ACCESS_STORAGE_KEY, JSON.stringify(normalizedSettings));
    window.dispatchEvent(new Event(ROLE_ACCESS_CHANGED_EVENT));
  }

  return normalizedSettings;
}

export function getAllowedModuleHrefs(
  role?: Pick<User, "role">["role"] | null,
  settings: RoleAccessSettings = getDefaultRoleAccessSettings()
) {
  if (!role) return new Set<string>();

  return new Set(normalizeRoleAccessSettings(settings)[role] ?? []);
}

export function canRoleAccessPath(
  role: Pick<User, "role">["role"] | null | undefined,
  pathname: string,
  settings: RoleAccessSettings = getDefaultRoleAccessSettings()
) {
  if (PUBLIC_PATHS.has(pathname)) return true;
  if (!role) return false;

  const matchedModule = allModuleHrefs
    .filter((href) => pathname === href || pathname.startsWith(`${href}/`))
    .sort((a, b) => b.length - a.length)[0];

  if (!matchedModule) return true;
  return getAllowedModuleHrefs(role, settings).has(matchedModule);
}

export function getRoleLandingPath(
  role: Pick<User, "role">["role"] | null | undefined,
  settings: RoleAccessSettings = getDefaultRoleAccessSettings()
) {
  if (!role) return "/login";

  const allowedHrefs = getAllowedModuleHrefs(role, settings);
  if (allowedHrefs.has("/dashboard")) return "/dashboard";

  return allModuleHrefs.find((href) => allowedHrefs.has(href)) ?? "/pengaturan";
}

export function canRoleReadCollection(
  role: Pick<User, "role">["role"] | null | undefined,
  collection: string
) {
  if (!role) return false;
  if (role === "admin") return true;

  const baseCollection = collection.startsWith("documents?category=") ? "documents" : collection;
  return READABLE_COLLECTION_ROLES[baseCollection]?.includes(role) ?? false;
}
